system/core/tags.php

```diff
@@ -0,0 +1,192 @@
+			/** @var HyphaDomElement $tagLang */
+			$id = $tagLang->parent()->getId();
+			$tag = ['label' => $tagLang->getAttribute('label'), 'description' => $tagLang->nodeValue];
+			if (in_array($id, $selectedTagIds)) $selectedTags[$id] = $tag;
+			else $deselectedTags[$id] = $tag;
+		}
+
+		// Construct html output
+		$html = '';
+		if (!empty($selectedTagIds) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div class="selectedTags">';
+			foreach ($selectedTags as $id => $tag) {
+				$html.= '<div class="tagSel_'.htmlspecialchars($id).'" class="tag">';
+				$html.= '<span title="'.htmlspecialchars($tag['description']).'">'.htmlspecialchars($tag['label']).'</span>';
+				if (isUser()) $html.= ' <span class="delete-tag" title="'.__('remove-tag').'" onclick="deselectTag('.htmlspecialchars(json_encode($id)).');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select class="remainingTags" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.htmlspecialchars($id).'" value="'.htmlspecialchars($id).'">'.htmlspecialchars($tag['label']).'</option>';
+				}
+				if (isAdmin()) $html.= '<option value="">'.__('new').'...</option>';
+				$html.= '</select>';
+			}
+		}
+		return $html;
+	}
+
+	/*
+		Function: hypha_indexTags
+		returns list of tags for the given page
+
+		Parameters:
+		$pageListNode - DOMElement containing page settings
+		$lang - language for tag index
+	*/
+	function hypha_indexTags($pageListNode, $lang) {
+		global $hyphaHtml;
+		$formId = $hyphaHtml->getDefaultForm()->getId();
+
+		if (!$pageListNode || !$lang) return '';
+
+		$language = hypha_pageGetLanguage($pageListNode, $lang);
+		$url = $language->getAttribute('id').'/'.$language->getAttribute('name');
+
+		ob_start();
+?>
+<script>
+	function selectTag(id) {
+		if (!id) addTag();
+		else {
+			hypha(<?=json_encode($url)?>, 'tagSelect', id, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
+			document.getElementById('tagList').innerHTML = 'loading...';
+		}
+	}
+	function deselectTag(id) {
+		hypha(<?=json_encode($url)?>, 'tagDeselect', id, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
+		document.getElementById('tagList').innerHTML = 'loading...';
+	}
+	function addTag() {
+		var name = prompt(<?=json_encode(__('enter-label'))?>, '');
+		if (name) {
+			var description = prompt(<?=json_encode(__('enter-description'))?>, '');
+			if (description) {
+				var json = '{ "name" : "' + name + '", "description" : "' + description + '"}';
```

Comment:
This breaks encoding when name or descriptoin contains a double-quote. I would instead build a JS object and use [JSON.Stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) to encode that to JSON (alternatively, pass the object directly to the `hypha()` function and let that automatically JSON-encode any non-strings or objects it receives).